src/Lib.P2P/Multiplex/Muxer.cs

```diff
@@ -0,0 +1,340 @@
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+using MultiFormats;
+using Nito.AsyncEx;
+
+namespace Lib.P2P.Multiplex
+{
+    /// <summary>
+    ///   Supports multiple protocols over a single channel (stream).
+    /// </summary>
+    /// <remarks>
+    ///   See <see href="https://github.com/libp2p/mplex"/> for the spec.
+    /// </remarks>
+    public class Muxer
+    {
+        private static ILog _log = LogManager.GetLogger(typeof(Muxer));
+
+        /// <summary>
+        ///   The next stream ID to create.
+        /// </summary>
+        /// <value>
+        ///   The session initiator allocates even IDs and the session receiver allocates odd IDs.
+        /// </value>
+        public long NextStreamId { get; private set; } = 1000;
+
+        /// <summary>
+        ///   The signle channel to exchange protocol messages.
+        /// </summary>
+        /// <value>
+        ///   A <see cref="Stream"/> to exchange protocol messages.
+        /// </value>
+        public Stream Channel { get; set; }
+
+        /// <summary>
+        ///   The peer connection.
+        /// </summary>
+        /// <value>
+        ///   The peer connection that owns this muxer.
+        /// </value>
+        public PeerConnection Connection { get; set; }
+
+        /// <summary>
+        ///   Raised when the remote end creates a new stream.
+        /// </summary>
+        public event EventHandler<Substream> SubstreamCreated;
+
+        /// <summary>
+        ///   Raised when the remote end closes a stream.
+        /// </summary>
+        public event EventHandler<Substream> SubstreamClosed;
+
+        private readonly AsyncLock _channelWriteLock = new AsyncLock();
+
+        /// <summary>
+        ///   The substreams that are open.
+        /// </summary>
+        /// <value>
+        ///   The key is stream ID and the value is a <see cref="Substream"/>.
+        /// </value>
+        public ConcurrentDictionary<long, Substream> Substreams = new ConcurrentDictionary<long, Substream>();
+
+        /// <summary>
+        ///   Determines if the muxer is the initiator.
+        /// </summary>
+        /// <value>
+        ///   <b>true</b> if the muxer is the initiator.
+        /// </value>
+        /// <seealso cref="Receiver"/>
+        public bool Initiator
+        {
+            get => (NextStreamId & 1) == 0;
+            set
+            {
+                if (value != Initiator)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)